port.js

```diff
@@ -5,233 +5,213 @@ const utqueue = require('ut-queue');
 const portStreams = require('./pull');
 const timing = require('./timing');
 const merge = require('./merge');
-const errorsFactory = require('./errors');
-const util = require('util');
+const createErrors = require('./errors');
 const EventEmitter = require('events');
 
-function Port(params) {
-    EventEmitter.call(this);
-    this.log = {};
-    this.logFactory = (params && params.logFactory) || null;
-    this.bus = (params && params.bus) || null;
-    this.errors = errorsFactory((this.bus && this.bus.errors) || params.errors);
-
-    this.sendQueues = utqueue.queues();
-    this.receiveQueues = utqueue.queues();
-    this.counter = null;
-    this.streams = [];
-    // performance metrics
-    this.methodLatency = null;
-    this.portLatency = null;
-    this.activeExecCount = null;
-    this.activeDispatchCount = null;
-    this.activeSendCount = null;
-    this.activeReceiveCount = null;
-    this.msgSent = null;
-    this.msgReceived = null;
-    // codec related metrics
-    this.activeEncodeCount = null;
-    this.activeDecodeCount = null;
-    this.bytesSent = null;
-    this.bytesReceived = null;
-    // state properties
-    this.isReady = false;
-    this.isConnected = new Promise(resolve => {
-        this.resolveConnected = resolve;
-    });
-    this.state = 'stopped';
-}
-
-util.inherits(Port, EventEmitter);
-
-Port.prototype.timing = timing;
-Port.prototype.merge = merge;
-
-Port.prototype.init = function init() {
-    this.logFactory && (this.log = this.logFactory.createLog(this.config.logLevel, {name: this.config.id, context: this.config.type + ' port'}, this.config.log));
-
-    if (this.config.metrics !== false && this.bus && this.bus.config.implementation && this.bus.performance) {
-        let measurementName = this.config.metrics || this.config.id;
-        let tags = {
-            host: os.hostname(),
-            impl: this.bus.performance.config.id || this.bus.config.implementation,
-            pid: process.pid
+module.exports = class Port extends EventEmitter {
+    constructor({ utLog, utBus, utError, config } = {}) {
+        super();
+        this.log = {};
+        this.utLog = utLog;
+        this.bus = utBus;
+        this.errors = createErrors(utError);
+        this.config = this.traverse(obj => obj.hasOwnProperty('defaults') && (obj.defaults instanceof Function ? obj.defaults.apply(this) : obj.defaults), config);
+        this.methods = this.traverse(obj => obj.hasOwnProperty('handlers') && (obj.handlers instanceof Function ? obj.handlers.apply(this) : obj.handlers), {});
```

Comment:
It might be a good idea to modify the signature of this.traverse to accept just the key as a first argument. The extraction function can be defined as a private variable there as it is always the same:
obj => obj.hasOwnProperty(key) && (obj[key] instanceof Function ? obj[key].call(this) : obj[key])
 
This way we'll reduce redundancy and the code here will become:
this.config = this.traverse('defaults', config);
this.methods = this.traverse('handlers', config);

Reply:
The idea of traverse is to be more generic and be used with other logic in the future. That's why this logic is outside.